File: Loenn/entities/TeleportToOtherSidePortal.lua
local drawableSprite = require("structs.drawable_sprite")

local TeleportToOtherSidePortal = {}

TeleportToOtherSidePortal.name = "XaphanHelper/TeleportToOtherSidePortal"
TeleportToOtherSidePortal.depth = 2000
TeleportToOtherSidePortal.fieldOrder = {
    "x", "y", "side", "warpSfx", "requireCassetteCollected", "requireCSideUnlocked", "flags", "teleportToStartingSpawnOfChapter", "wipeType", "wipeDuration", "registerCurrentSideAsCompelete"
}
TeleportToOtherSidePortal.canResize = {false, false}
TeleportToOtherSidePortal.fieldInformation = {
    side = {
        options = {"A-Side", "B-Side", "C-Side"},
        editable = false
    },
    wipeType = {
        options = {"Spotlight", "Curtain", "Mountain", "Dream", "Starfield", "Wind", "Drop", "Fall", "KeyDoor", "Angled", "Heart", "Fade"},
        editable = false
    }
}
TeleportToOtherSidePortal.placements = {
    name = "TeleportToOtherSidePortal",
    data = {
        side = "A-Side",
        requireCassetteCollected = false,
        requireCSideUnlocked = false,
        flags = "",
        wipeType = "Fade",
        wipeDuration = 1.35,
        warpSfx = "event:/game/xaphan/warp",
        teleportToStartingSpawnOfChapter = false,
        registerCurrentSideAsCompelete = false
    }
}

function TeleportToOtherSidePortal.sprite(room, entity)
    local texture = entity.side or "A-Side"
    local sprite = drawableSprite.fromTexture( "objects/XaphanHelper/TeleportToOtherSidePortal/" .. texture .."00", entity)

    return sprite
end

return TeleportToOtherSidePortal

File: Ahorn/entities/TeleportToOtherSidePortal.jl
module XaphanHelperTeleportToOtherSidePortal

using ..Ahorn, Maple

@mapdef Entity "XaphanHelper/TeleportToOtherSidePortal" TeleportToOtherSidePortal(x::Integer, y::Integer, side::String="A-Side", requireCassetteCollected::Bool=false, wipeType::String="Fade", wipeDuration::Number=1.35, warpSfx::String="event:/game/xaphan/warp", teleportToStartingSpawnOfChapter::Bool=false, requireCSideUnlocked::Bool=false, flags::String="", registerCurrentSideAsCompelete::Bool=false)

const placements = Ahorn.PlacementDict(
    "Teleport To Other Side Portal (Xaphan Helper)" => Ahorn.EntityPlacement(
        TeleportToOtherSidePortal,
        "point",
        ),
    )

    function Ahorn.selection(entity::TeleportToOtherSidePortal)
        x, y = Ahorn.position(entity)
        side = get(entity.data, "side", "A-Side")
        if side == ""
            side = "A-Side"
        end
        return Ahorn.getSpriteRectangle("objects/XaphanHelper/TeleportToOtherSidePortal/$(side)00.png", x, y)
    end

    Ahorn.editingOptions(entity::TeleportToOtherSidePortal) = Dict{String, Any}(
    "side" => String["A-Side", "B-Side","C-Side"],
    "wipeType" => String["Spotlight", "Curtain", "Mountain", "Dream", "Starfield", "Wind", "Drop", "Fall", "KeyDoor", "Angled", "Heart", "Fade"]
)

    function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::TeleportToOtherSidePortal, room::Maple.Room)
        side = get(entity.data, "side", "A-Side")
        if side == ""
            side = "A-Side"
        end
        Ahorn.drawSprite(ctx, "objects/XaphanHelper/TeleportToOtherSidePortal/$(side)00.png", 0, 0, jx=0.5, jy=0.5)
    end

end